modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java

```diff
@@ -4270,9 +4277,11 @@ private Checkpoint markCheckpointBegin(CheckpointMetricsTracker tracker) throws
 
             IgniteFuture snapFut = null;
 
-            IgniteBiTuple<Collection<GridMultiCollectionWrapper<FullPageId>>, Integer> cpPagesTuple;
+            T3<Collection<T2<PageMemoryEx, GridMultiCollectionWrapper<FullPageId>>>, Integer, Boolean> cpPagesTriple;
```

Comment:
Maybe it makes sense to replace this ugly T3 + T2 construction to some specific class with clear fields?